include/secp256k1_schnorrsig.h

```diff
@@ -50,59 +52,125 @@ typedef int (*secp256k1_nonce_function_hardened)(
+ *                this, it is only supplemental to security and can be NULL. See
+ *                BIP-340 for a full explanation of this argument and for
+ *                guidance if randomness is expensive.
  */
 SECP256K1_API int secp256k1_schnorrsig_sign(
     const secp256k1_context* ctx,
     unsigned char *sig64,
-    const unsigned char *msg32,
+    const unsigned char *msg,
+    size_t msg_len,
     const secp256k1_keypair *keypair,
+    unsigned char *aux_rand32
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(5);
+
+/** Create a schnorrsig config (in dynamically allocated memory).
+ *
+ *  This function uses malloc to allocate memory. It is guaranteed that malloc is
+ *  called at most once for every call of this function.
+ *
+ *  The allocated memory must be freed with secp256k1_schnorrsig_config_destroy.
+ *
+ *  Returns: a newly created schnorrsig config
+ *  Args: ctx:  an existing context object (cannot be NULL)
+ */
+SECP256K1_API secp256k1_schnorrsig_config* secp256k1_schnorrsig_config_create(
+    const secp256k1_context* ctx
+) SECP256K1_ARG_NONNULL(1);
+
+/** Destroy a schnorrsig config (created in dynamically allocated memory).
+ *
+ *  The config pointer may not be used afterwards.
+ *  Args:  ctx: a secp256k1 context object
+ *  In: config: the config to destroy
+ */
+SECP256K1_API void secp256k1_schnorrsig_config_destroy(
+    const secp256k1_context* ctx,
+    secp256k1_schnorrsig_config *config
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2);
+
+/** Set nonce function pointer and nonce data of a config object
+ *
+ *  Returns: 1 if the arguments are valid. 0 otherwise
+ *  Args:  ctx: a secp256k1 context object
+ *  In: config: the config to set the noncefp and ndata for
+ *     noncefp: the nonce function pointer to set
+ *       ndata: the nonce data to set
+ */
+SECP256K1_API int secp256k1_schnorrsig_config_set_nonce(
+    const secp256k1_context* ctx,
+    secp256k1_schnorrsig_config *config,
     secp256k1_nonce_function_hardened noncefp,
     void *ndata
-) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2) SECP256K1_ARG_NONNULL(3) SECP256K1_ARG_NONNULL(4);
+) SECP256K1_ARG_NONNULL(1) SECP256K1_ARG_NONNULL(2);
+
+/** Create a Schnorr signature with a more flexible API.
+ *
+ *  Same arguments as secp256k1_schnorrsig_sign except that it accepts a pointer
+ *  to a config object that allows customizing signing by passing additional
+ *  arguments.
+ *
+ *  In: noncefp: pointer to a nonce generation function. If NULL,
+ *               secp256k1_nonce_function_bip340 is used
+ *        ndata: pointer to arbitrary data used by the nonce generation function
+ *               (can be NULL). If it is non-NULL and
+ *               secp256k1_nonce_function_bip340 is used, then ndata must be a
+ *               pointer to 32-byte auxiliary randomness as per BIP-340.
+ *       config: pointer to a config object.
+ */
```

Comment:
Perhaps I am misunderstanding the doc comment syntax but something is not quite right here. It should say something like "instead of `aux_rand32` it accepts a custom config...". I think then the only "In" should be the config?

Reply:
You're right, the doc for `ndata` and `noncefp` here should be removed in the last commit.

Reply:
fixed